### DISCUSSION TITLE
lokinet v0.9.4

### DISCUSSION THREAD

**Discussion by majestrate** [oxen-io/lokinet#1683] [General]
# lokinet 0.9.4

this release is a mandatory upgrade for service nodes, highly recommended for clients.

this release will have an accompanying release of storage server and oxend.

## critical bug fixes (service nodes and clients)

* resolve inter node communication bug where multiple parallel outbound session attempts were being sent to the same remote udp endpoint causing a state race condition where establishing a new outbound session times out.

## new client features
* new config option `[network]:persist-addrmap-file` to persist ephemeral ip mappings for inbound flows so that between restarts ip addresses dont remap for active flows, this allows tcp connections to stick between restarts.
* new config option `[dns]:add-hosts` to add a dns hosts file by path for client side dns filtration, i.e. adblock and such.
* allow non-standard dns port in upstream dns

## client side updates
* add upper bound on parallel ONS lookups per name
* refactor some outbound snapp flow code to make it functional, total rewrite of that layer will happen soon after tag (probably).
* redo snapp flow handover logic

## build system / ci

* static deps now are built in parallel 
* removed old android cruft, replaced gradle build system with `./contrib/android.sh`
* added flutter to android ci docker image, removed gradle from android docker ci image
* fixed typo in debian-stable ci docker image
* package jni libs in android ci pipeline
* package lokinet-bootstrap binary instead of the shell script in ci pipeline

## android

* add `DumpStatus` native method to `network.loki.lokinet.LokinetDaemon` for state introspection

## misc / utils

* added `contrib/bencode-dump.py` script for printing out bencoded files
* `lokinetmon` updates to match changes in RPC (removal of `badIntros` field)
* updates to readme.md about building from source
* we will no longer provide static linux binaries in releases, users of oxen-rancher (the only users that should even care about these to begin with) should use CI builds from the [stable branch](https://oxen.rocks/oxen-io/loki-network/lokinet-stable-linux-amd64-LATEST.tar.xz)

<hr /><em>This discussion was created from the release <a href='https://github.com/oxen-io/loki-network/releases/tag/v0.9.4'>lokinet v0.9.4</a>.</em>

**Comment by jb-gf** [comment]
Red car is now available we fed up with your house right after the driveway
